Getting Started with OpenTelemetry Rust Logs in 5 Minutes - Exporting logs to user_events (#2)
lalitb/opentelemetry-rust: opentelemetry-user-events-logs/examples/README.md

# Getting Started with OpenTelemetry Rust Logs in 5 Minutes - Exporting logs to user_events

## Prerequisites

* Rust and Cargo: Install them if you haven't.

> * A Linux system: Make sure it supports user_events.

Reviewer: How to "make sure"?

Author: Have added more details - 

A Linux system with [user_events](https://docs.kernel.org/trace/user_events.html) enabled: Ubuntu 23.10 (once available), or kernel 6.4 or later build with user_events support.

* Correct Permissions: Linux user(preferable non-root) with write permission to the tracefs.

> * A listener/agent to listen for new events. Or else `perf` and `decode-perf` tools installed for validation.

Reviewer: Provide official links to these prerequisites.

Author: Have added them under the useful links section at the end. Will put it here instead.

## Steps
1. Create a new Rust application and run it.

```sh
$ cargo new otel-userevents-getting-started
$ cd otel-userevents-getting-started
$ cargo run
```

You'll see "Hello, world!".

> 2. Open the `Cargo.toml` file and paste these lines

Reviewer: Is this what typical users would do during development?

Author: Yes, that's the correct approach to add dependencies for the application.

```toml
[dependencies]
opentelemetry_sdk = "0.20.0"
opentelemetry-appender-tracing = "0.1.0"
opentelemetry-user-events-logs = "0.1.0"
tracing = { version = "0.1.37", default-features = false, features = ["std"] }
tracing-core = "0.1.31"
tracing-subscriber = { version = "0.3.0", default-features = false, features = ["registry", "std"] }

```

3. Replace what's inside `src/main.rs` with the example file [basic.rs](./basic.rs). This new code sets up logging, make new tracepoints, and send an error message where it needs to go.
4. Run and Check: Go back to the terminal and type `cargo run`. If everything is okay, your logs will be sent to the proper place. To test this, you can use the `perf` and `decode-perf` tool like this:

```sh
$ perf record -e user_events:test_L2K1Gtest
<ctrl-c after running sample code>

> $ decode-perf ./perf.data

Reviewer 2: pre-requisite missing to get decode-perf?

Reviewer 2: could you include a sample output json in this readme itself?

Author: > pre-requisite missing to get decode-perf?

It's there :)

Reviewer 2: Can't find anything pointing to `LinuxTracepoints` in pre-req....

Author: It's under the "Useful links" section at the end, but I can see the confusion. Will put it inside the pre-requisite.

Author: Have update this in pre-requisite now.

<log data is output here>
```

> Note: The sample code sets up a tracepoint called `test_L2K1Gtest`, and error message are relayed to this tracepoint. Refer to next section for more details.

Reviewer: 1. Change "sample" to "example" ("sample" has many different meanings in the context of OpenTelemetry).
2. What does "error message" and "relay" mean?

Author: Changed to "error log is exported to this tracepoint", along with other comments.

## Configuring tracepoint names:

> ### Tracepoint Format:
> ProviderName + '_' + 'L' + EventLevel + 'K' + EventKeyword + 'G' + ProviderGroup

Reviewer: ```suggestion
### Tracepoint Format:

ProviderName + '_' + 'L' + EventLevel + 'K' + EventKeyword + 'G' + ProviderGroup
```

### What you need to know:

* `ProviderName` and `EventKeyword` can be changed using Exporter configuration.
* These settings decide the tracepoint names and where events go.

> * OpenTelemetry SDK will automatically create a tracepoint for each of the 5 levels (Critical, Error, Warning, Info, Verbose).

Reviewer 2: ```suggestion
* OpenTelemetry SDK currently creates a tracepoint for each of the 5 levels (Critical, Error, Warning, Info, Verbose).
```

* In future, the `EventKeyword` would be mapped to the unique events sent by the application.

### How to configure:

> 1. Create ExporterConfig with below settings:

Reviewer 2: I think 2,3 here are irrelevant, and we should hide it from users.

Author: step 2 is setting up the provider_name. Will remove step 3.

Reviewer 2: yes. but it can be done later. 
something like loggerProviderBuilder.add_user_event_exporter(provider_name)

```rust
    let exporter_config = ExporterConfig {
        default_keyword: 1,
        keywords_map: HashMap::new(),
    };
```
`EventKeyword` is configured with `default_keyword` value above.

2. Create LogProcessor instance:
```rust

> let processor = ReentrantLogProcessor::new("test_provider", None, exporter_config);

Reviewer: Curious - what's the semantic of the first parameter?

Author: It's documented here - https://github.com/microsoft/LinuxTracepoints/tree/main/libeventheader-tracepoint#provider-names :

--clip
A provider is a component that generates events. Each event from a provider is associated with a Provider Name that uniquely identifies the provider.

The provider name should be short, yet descriptive enough to minimize the chance of collision and to help developers track down the component generating the events. Hierarchical namespaces may be useful for provider names, e.g. `MyCompany_MyOrg_MyComponent`.
-clap

Have added the link in this document.

Author: So a Provider can generate different unique events.  In future, we can have each event mapped to a unique EventKeyword. And the combination of `ProviderName + EventKeyword + Level` is eventually mapped to a tracepoint name.

Reviewer: I don't understand why a LinuxTracepoint provider has something to do with a ReentrantLogProcessor. If you look at the name of `ReentrantLogProcessor`, it has nothing to do with provider.

Author: Yes, the plan is to move the provider configuration into ExporterConfig for the next release. I will be raising the PR for the same.

```
The `ProviderName` is configured as `test_provider` above.

3. Create LoggerProvider with above configurations.
```rust
let provider =  LoggerProvider::builder().with_log_processor(reenterant_processor).build();
```

### Sample code:

> The sample code uses `test` as the ProviderName and `1` as the EventKeyword. So, it makes these 5 tracepoints:

Reviewer 2: "myprovider" instead of "test" to make smoother/easier discussions :)

Reviewer: +1

Author: renamed to "myprovider"

# Getting Started with OpenTelemetry Rust Logs in 5 Minutes - Exporting logs to user_events

## Prerequisites

* Rust and Cargo: [Install](https://doc.rust-lang.org/cargo/getting-started/installation.html) them if you haven't.
* A Linux system with [user_events](https://docs.kernel.org/trace/user_events.html) enabled: Ubuntu 23.10 (once available), or kernel 6.4 or later build with user_events support.

> * Correct Permissions: Linux user(preferable non-root) with write permission to the tracefs.

Reviewer: ```suggestion
* Correct Permissions: Linux user (preferable non-root) with write permission to the tracefs.
```

> * A listener/agent to listen for new events. Or else [perf](https://perf.wiki.kernel.org/index.php/Main_Page) and [decode-perf](https://github.com/microsoft/LinuxTracepoints/tree/main/libeventheader-decode-cpp/tools) tools installed for validation.

Reviewer: What does "or else" mean?

Reviewer: Since this is more of a getting-started example, we don't have to list all possible ways, I suggest that we just provide the most common way or an easy way (e.g. download XYZ tool and use that).

Author: Good point. One documenting the perf/decode-perf tool now.

> * Decode-Perf Tool: Install the [decode-perf](https://perf.wiki.kernel.org/index.php/Main_Page) tool to decode the user_events.

Reviewer 2: wrong link for decode-perf tool.

Author: Yeah, my bad :( Thanks for noticing this.

## Steps
1. Create a new Rust application and run it.

```sh
$ cargo new otel-userevents-getting-started
$ cd otel-userevents-getting-started
$ cargo run
```

You'll see "Hello, world!".

2. Open the `Cargo.toml` file and paste these lines
```toml
[dependencies]
opentelemetry_sdk = "0.20.0"
opentelemetry-appender-tracing = "0.1.0"
opentelemetry-user-events-logs = "0.1.0"
tracing = { version = "0.1.37", default-features = false, features = ["std"] }
tracing-core = "0.1.31"
tracing-subscriber = { version = "0.3.0", default-features = false, features = ["registry", "std"] }

```

3. Replace what's inside `src/main.rs` with the example file [basic.rs](./basic.rs). This new code sets up logging, make new tracepoints, and send an error log where it needs to go.
4. Run and Check: Go back to the terminal and type `cargo run`. If everything is okay, your logs will be sent to the proper place. To test this, you can use the `perf` and `decode-perf` tool like this:

```sh
$ perf record -e user_events:myprovider_L2K1Gmyprovider
<ctrl-c after running example code>
$ decode-perf ./perf.data

> <log data is output here>

Reviewer 2: can you paste an actual output, so it is easy to compare where things land?

Author: Sure, added now.

```

Note: The example code sets up a tracepoint called `myprovider_L2K1Gmyprovider`, and error log is exported to this tracepoint. Refer to the next section for more details.

## Configuring tracepoint names:

### Tracepoint Format:

ProviderName + '_' + 'L' + EventLevel + 'K' + EventKeyword + 'G' + ProviderGroup

### What you need to know:

* [ProviderName](https://github.com/microsoft/LinuxTracepoints/tree/main/libeventheader-tracepoint#provider-names) and [EventKeyword](https://github.com/microsoft/LinuxTracepoints/tree/main/libeventheader-tracepoint#tracepoint-names) can be changed using Exporter configuration.

> * `ProviderGroup` is assigned the value of `ProviderName`.

Reviewer 2: does the user need to know this?

Author: I thought it is good for them to know that `ProviderGroup`is not configurable, and also make them understand why `myprovider` comes twice in tracepoint name `myprovider_L5K1Gmyprovider`

Reviewer 2: Got it.! Yea it can be confusing without this to see myprovider twice.

* In future, the `EventKeyword` would be mapped to the unique events sent by the application.

### How to configure:

1. Create ExporterConfig with below settings:
```rust
    let exporter_config = ExporterConfig {
        default_keyword: 1,
        keywords_map: HashMap::new(),
    };
```
`EventKeyword` is configured with `default_keyword` value above.

2. Create LogProcessor instance:
```rust
let processor = ReentrantLogProcessor::new("myprovider", None, exporter_config);
```
The `ProviderName` is configured as `myprovider` above.

3. Create LoggerProvider with above configurations.
```rust
let provider =  LoggerProvider::builder().with_log_processor(reenterant_processor).build();
```

### Example code:

The example code uses `test` as the ProviderName and `1` as the EventKeyword. So, it makes these 5 tracepoints:

> * test_L5K1Gtest

Reviewer: This needs to be updated to reflect the provider name change?

* test_L4K1Gtest
* test_L3K1Gtest
* test_L2K1Gtest
* test_L1K1Gtest

To get the events sent to these tracepoints, the listener should be set up to monitor one or more of them.

> 
> 

Reviewer: ```suggestion

```

> Useful Links:

Reviewer: ```suggestion
## Useful links
```
